### PR TITLE
Add histogram exemplar support

### LINKED ISSUE #119697: Trace exemplars for kube-apiserver metrics
### What would you like to be added?

Add exemplars to APIServer metrics.  To accomplish this, we need to add FooWithExemplar functions, like histograms ObserveWithExemplar to component-base/metrics: 
https://github.com/prometheus/client_golang/blob/7f2db5f1ed43d8a4312bfd85280a17dfef13358b/prometheus/histogram.go#L722 At minimum, we should allow exemplars on counters and histograms.

Then, we need to change Observe() calls to ObserveWithExemplar calls, similar to https://github.com/prometheus/client_golang/blob/main/examples/exemplars/main.go.

We should use the `trace_id` key for the trace ID, and the `span_id` key for the span ID.

cc @Richabanker @kubernetes/sig-instrumentation-leads 

### Why is this needed?

With https://github.com/kubernetes/enhancements/issues/647, we can now collect traces for requests to the kubernetes api.  We have a number of metrics, such as `apiserver_request_duration_seconds`, which would benefit from having exemplars attached.

### PR DESCRIPTION
**Pull request by richabanker**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- Adds exemplar support for histogram
- Adds exemplar support for histogramVec
- Rename exemplarMetric to exemplarCounterMetric and move it to counter.go for better code organization

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/119697

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-base/metrics/metric.go [hunk 1]
```diff
@@ -221,15 +221,16 @@ var noop = &noopMetric{}
...
-func (noopMetric) Write(*dto.Metric) error           { return nil }
-func (noopMetric) Describe(chan<- *prometheus.Desc)  {}
-func (noopMetric) Collect(chan<- prometheus.Metric)  {}
+func (noopMetric) Inc()                                           {}
+func (noopMetric) Add(float64)                                    {}
+func (noopMetric) Dec()                                           {}
+func (noopMetric) Set(float64)                                    {}
+func (noopMetric) Sub(float64)                                    {}
+func (noopMetric) Observe(float64)                                {}
+func (noopMetric) ObserveWithExemplar(float64, prometheus.Labels) {}
```
Comment (rexagod) [inline staging/src/k8s.io/component-base/metrics/metric.go:230]: I'm revisiting this after a [good while](https://github.com/prometheus/prometheus/pull/12738) but IIRC from a discussion with @logicalhan around exemplar support for counters, we want to modify the [existing methods](https://github.com/kubernetes/kubernetes/pull/119949/files#diff-e318d079505983653611023d8593be867ebd92d90f9522093188b25dbeca95a3R103) implicitly (without modifying their signatures) to support exemplars OOTB.

Additionally, my understanding is that we don't want to make any component [directly dependent](https://github.com/kubernetes/kubernetes/pull/124013/files#diff-c4f07ee44a4626c817dd95cf30e147a2125ea923804be5c3608fb4a845fc1c20R31) on `prometheus/client_golang`, but incorporate that behavior into `component-base` and have them import that instead, in an effort to be able to offset any breaking changes and not having to send patches upstream (and catering to any extra use-cases specific to `k/k`).

All components themselves will be responsible for ["attaching" exemplars within their contexts](https://github.com/kubernetes/kubernetes/pull/124013/files#diff-c4f07ee44a4626c817dd95cf30e147a2125ea923804be5c3608fb4a845fc1c20R31), if they want to utilize this feature-set, but we won't have to make any explicit changes in any component for this feature-set to work. However, the SIG can help components benefit from this by patching their contexts for them, but I believe that is out of the scope of this effort (adding exemplar support to `component-base`).

Comment (richabanker) [reply staging/src/k8s.io/component-base/metrics/metric.go]: Thanks for the pointers @rexagod. The initial version of this PR was mostly a POC to verify that we are able to see exemplars after-all in metrics, hence admittedly had quite a lot of hacky stuff going on at first.

I modified this PR now, following what you are doing for counters in https://github.com/kubernetes/kubernetes/pull/119949, though it still needs to use the `component-base.metrics.exemplarMetric` that you have introduced, so I'll wait for that PR to merge first and make those changes then.

Additionally I am wondering, if we need to introduce Add()/Observe() methods for CounterVec/HistogramVec too?

**Review by logicalhan — COMMENTED** [review]
/assign
/triage accepted

File: staging/src/k8s.io/component-base/metrics/histogram.go [hunk 2]
```diff
@@ -21,17 +21,45 @@ import (
...
+
+func (h *Histogram) Observe(v float64) {
+	h.withExemplar(v)
+}
+
+func (h *Histogram) withExemplar(v float64) {
+	(&exemplarHistogram{h}).withExemplar(v)
+}
+
+func (h *exemplarHistogram) withExemplar(v float64) {
```
Comment (dashpole) [inline staging/src/k8s.io/component-base/metrics/histogram.go:49]: Why do you need the extra `exemplarHistogram` type and the withExemplar functions.

**Review by dashpole — COMMENTED** [review]
lgtm with one nit

File: staging/src/k8s.io/component-base/metrics/histogram.go (hunk 2 above)
Comment (richabanker) [reply staging/src/k8s.io/component-base/metrics/histogram.go]: Good catch, I don't. Can't remember why I did that, removed it now.

File: staging/src/k8s.io/component-base/metrics/histogram.go [hunk 3]
```diff
@@ -21,17 +21,37 @@ import (
...
+func (h *Histogram) withExemplar(v float64) {
+	var exemplarLabels prometheus.Labels
+	maybeSpanCtx := trace.SpanContextFromContext(h.ctx)
+	if maybeSpanCtx.IsValid() {
+		exemplarLabels = prometheus.Labels{
+			"trace_id": maybeSpanCtx.TraceID().String(),
+			"span_id":  maybeSpanCtx.SpanID().String(),
+		}
+	}
+	if m, ok := h.ObserverMetric.(prometheus.ExemplarObserver); ok {
```
Comment (dashpole) [inline staging/src/k8s.io/component-base/metrics/histogram.go:50]: is this always going to be true?  Should we fall back to the regular observe if it isn't an examplar observer?

File: staging/src/k8s.io/component-base/metrics/histogram.go [hunk 4]
```diff
@@ -21,17 +21,37 @@ import (
...
 	lazyMetric
 	selfCollector
 }
 
+func (h *Histogram) Observe(v float64) {
+	h.withExemplar(v)
+}
+
+func (h *Histogram) withExemplar(v float64) {
+	var exemplarLabels prometheus.Labels
```
Comment (dashpole) [inline staging/src/k8s.io/component-base/metrics/histogram.go:42]: Should we move this inside the `if` statement below to avoid unnecessary computation?

**Review by dashpole — COMMENTED** [review]
same comments for exemplarHistogramVec

**Comment by richabanker** [comment]
> same comments for exemplarHistogramVec

I need to wrap the return value of `func (vc *HistogramVecWithContext) WithLabelValues(lvs ...string)` which is [currently](https://github.com/kubernetes/kubernetes/blob/release-1.30/staging/src/k8s.io/component-base/metrics/histogram.go#L207) `ObserverMetric` into a struct on which I could apply `func withExemplar()`. So the extra `exemplarHistogramVec` is needed for exemplarHistogramVec.

File: staging/src/k8s.io/component-base/metrics/histogram.go (hunk 3 above)
Comment (richabanker) [reply staging/src/k8s.io/component-base/metrics/histogram.go:42]: I am not sure... but still, added a fallback to use the regular Observe() flow.

File: staging/src/k8s.io/component-base/metrics/histogram.go (hunk 4 above)
Comment (richabanker) [reply staging/src/k8s.io/component-base/metrics/histogram.go]: Sounds good, moved.

**Review by dashpole — APPROVED** [review]
/lgtm

**Pull request merged**